src/js/conntest.js

```diff
@@ -40,32 +40,73 @@ function hostConnectivityTest() {
 }
 
 function runConnectivityTest(iceCandidateFilter, config) {
+  var parsedCandidates = [];
   var call = new Call(config);
   call.setIceCandidateFilter(iceCandidateFilter);
+
+  // Collect all candidate for validation.
+  call.pc1.onicecandidate = function(event) {
+    if (event.candidate) {
+      var parsedCandidate = Call.parseCandidate(event.candidate.candidate);
+      parsedCandidates.push(parsedCandidate);
+
+      // Report candidate info based on iceCandidateFilter.
+      if (iceCandidateFilter(parsedCandidate)) {
+        reportInfo(
+          'Gathered candidate of Type: ' + parsedCandidate.type +
+          ' Protocol: ' + parsedCandidate.protocol +
+          ' Address: ' + parsedCandidate.address);
+      }
+    }
+  };
+
   var ch1 = call.pc1.createDataChannel(null);
   ch1.addEventListener('open', function() {
     ch1.send('hello');
   });
   ch1.addEventListener('message', function(event) {
-    clearTimeout(timeout);
     if (event.data !== 'world') {
-      reportFatal();
+      reportError('Invalid data transmitted.');
     } else {
       reportSuccess('Data successfully transmitted between peers.');
-      setTestFinished();
     }
+    hangup();
   });
   call.pc2.addEventListener('datachannel', function(event) {
     var ch2 = event.channel;
     ch2.addEventListener('message', function(event) {
       if (event.data !== 'hello') {
-        clearTimeout(timeout);
-        reportFatal();
+        hangup('Invalid data transmitted.');
       } else {
         ch2.send('world');
       }
     });
   });
   call.establishConnection();
-  timeout = setTimeout(reportFatal.bind(null, 'Timed out'), 2000);
+  timeout = setTimeout(hangup.bind(null, 'Timed out'), 5000);
+
+  function findParsedCandidateOfSpecifiedType(candidateTypeMethod) {
+    for (var candidate in parsedCandidates) {
+      if (candidateTypeMethod(parsedCandidates[candidate])) {
+        return candidateTypeMethod(parsedCandidates[candidate]);
+      }
+    }
+  }
+
+  function hangup(errorMessage) {
```

Comment:
I am not sure if call.close()/pc.close() does stop the peer connection immediatly and flushes away any pending ondata event. If it does not there is a chance this hangup is called twice... gosh we really need to start passing a test object instead of having this global methods to reportError/Warning/etc...

I guess no action needs to be done now, but we should fix it eventually.